File: Classes/spring_introduction/src/main/java/com/example/spring_introduction/repository/MemoryMemberRepository.java
package com.example.spring_introduction.repository;

import com.example.spring_introduction.domain.Member;

import java.util.*;

public class MemoryMemberRepository implements MemberRepository {

    private static Map<Long, Member> store = new HashMap<>();
    private static long sequence = 0L;

    @Override
    public Member save(Member member) {
        member.setId(++sequence);
        store.put(member.getId(), member);

        return member;
    }

    @Override
    public Optional<Member> findById(Long id) {
        return Optional.ofNullable(store.get(id));
    }

    @Override
    public Optional<Member> findByName(String name) {
        return store.values().stream()
                .filter(member -> member.getName().equals(name))
                .findAny(); // @NotNull Optional<T> findAny();
    }

    @Override
    public List<Member> findAll() {
        return new ArrayList<>(store.values());
    }

    /**
     * 테스트 코드에서 단위테스트 이후 원복을 위한 매서드
     */
    public void clearStore() {
        store.clear();
    }
}

File: Classes/spring_basic/src/main/kotlin/com/example/spring_basic/member/MemoryMemberRepository.kt
package com.example.spring_basic.member

import org.springframework.stereotype.Component

@Component
class MemoryMemberRepository : MemberRepository {

    companion object {
        private val store: MutableMap<Long, Member> = HashMap()
    }

    override fun save(member: Member) {
        store[member.id] = member
    }

    override fun findById(memberId: Long): Member {
        return store[memberId]!!
    }
}